### PR TITLE
[BLOG] The Convoluted Story Behind `np.top_k`

### PR DESCRIPTION
**Pull request by JuliaPoo**
## Text styling

- [x] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [x] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [x] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [x] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [x] Blog post featured image is in PNG or JPEG format, **not** SVG.
- [x] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [x] If there are emojis, there are not more than three in a row.
- [x] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [x] If it were to be read as plain text, the blog still makes sense and no information is missing.

### CODE REVIEW

File: apps/labs/posts/convoluted-story-behind-topk.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+hero:
+  imageSrc: /posts/convoluted-story-behind-topk/NumPy_logo_2020.webp
+  imageAlt: 'NumPy Logo retrieved from Wikipedia Commons'
+---
+
+In this blog post, I describe my experience as a first-time contributor and talk about the story behind `np.top_k`.
+
+---
+
+Hi~ I'm [Jules Poon](https://juliapoo.github.io/), a Year 2 undergraduate at the National University of Singapore. This summer I had the incredible opportunity to join the Open Source internship at Quansight Labs, under the mentorship of [Ralf Gommers](https://github.com/rgommers), [Nathan Goldbaum](https://ngoldbaum.github.io/) and [Matti Picus](https://github.com/mattip) for work on NumPy, and [Evgeni Burovski](https://github.com/ev-br) for work
```
Comment (rgommers) [inline apps/labs/posts/convoluted-story-behind-topk.md:19]: could you add "program" after "internship"?

Also, that home page of Nathan is pretty outdated - probably better to link https://github.com/ngoldbaum

File: apps/labs/posts/convoluted-story-behind-topk.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+on SciPy.
+
+My task for NumPy is quite generic: To contribute to NumPy. So I'm placed in a position most people just beginning in open source will find themselves in. Hopefully, this blog post can help somebody just starting with open-source to navigate projects and understand the process behind the decisions made in open-source projects.
```
Comment (rgommers) [inline apps/labs/posts/convoluted-story-behind-topk.md:22]: Re position: probably worth pointing out that there are two important differences: you got recommendations of good starting issues, and you had direct access to maintainers for discussion - both of which make it easier compared to starting out as a new contributor alone.

File: apps/labs/posts/convoluted-story-behind-topk.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+Lastly, to catch up on existing discussions or to create new ones, in addition to the repository's issues and PR pages, one can look through (and join) the [Mailing List](https://mail.python.org/mailman3/lists/numpy-discussion.python.org/). The Mailing List also announces the details of the biweekly community meetings where you can ask questions and discuss details.
+
+## My Contributions
+
+Throughout the 1.5 months working on NumPy, I made incremental improvements to the tests, documentation, CI, and submitted bug fixes. I also worked on furthering the discussion for `np.top_k` and drafted implementations, and I want to talk more about the process behind that.
+
+### Reviewing Community Discussions
+
+The discussions for `np.top_k` started from [gh-15128 Feature request: Allow np.argmax to output top K maximum values](https://github.com/numpy/numpy/issues/15128) in 2019 where multiple people expressed their request for having this feature in NumPy.
```
Comment (rgommers) [inline apps/labs/posts/convoluted-story-behind-topk.md:42]: Perhaps worth stating that it's the 6th-highest feature request in the whole issue tracker as measured by 👍🏼 emoji's? (that's how we looked for impactful topics)

File: apps/labs/posts/convoluted-story-behind-topk.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    to be the earliest/latest occurrence of the element. E.g.,
+    ``np.top_k([3,3,3], 1)`` can return ``(array([3]), array([1]))``
+    rather than ``(array([3]), array([0]))`` or
+    ``(array([3]), array([2]))``.
+
+    ...
+    """
+```
+
+I decided on `np.top_k` as the name of the function, and against an additional function like `argtop_k` as there isn't a significant performance difference for just returning both the indices and values in one call (which Ralf agrees).
```
Comment (rgommers) [inline apps/labs/posts/convoluted-story-behind-topk.md:118]: minor textual "(which Ralf agreed with)"

File: apps/labs/posts/convoluted-story-behind-topk.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+I made heavy reference to the [array-api PR](https://github.com/data-apis/array-api/pull/722) and the notes there as well. Something I changed was to have `axis=-1` by default instead of the proposed `axis=None`. This follows `np.argparitition` and my own sensibility.
```
Comment (rgommers) [inline apps/labs/posts/convoluted-story-behind-topk.md:120]: typo in `argpartition`

File: apps/labs/posts/convoluted-story-behind-topk.md [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+> Q: Should we be more strict in specifying what should happen when k exceeds the number of elements?
+>
+> A: `np.argpartition` does raise a `ValueError` when that happens, so yes.
+
+> Q: How should `np.top_k` handle `NaN` values?
+>
+> A: `np.argpartition` (at this point in time) does not define the sort order of `np.nan`, even though `np.sort` does, so `np.top_k` shouldn't either.
+
+The behaviour for `NaN` values will become a huge contention later.
```
Comment (rgommers) [inline apps/labs/posts/convoluted-story-behind-topk.md:132]:
```suggestion
The behaviour for `NaN` values will become a huge point of contention later.
```

File: apps/labs/posts/convoluted-story-behind-topk.md [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+### Feedback for the First Implementation
+
+The performance of the implementation is good, being significantly faster than a full `np.sort`. However, [Jake Vanderplas](https://github.com/jakevdp) insists that `np.top_k` should clarify the behaviour of `NaN`s and not just leave it unspecified. [Sebastian Berg](https://github.com/seberg) pointed out two options on how to approach this:
+
+1. Leave the sort order of `np.nan` unchanged (that it's the biggest element).
+2. Always sort `np.nan` to the end.
+
+The significance of Option 2 is that `np.top_k([1., np.nan], 1)` will return `1.` instead of `nan`, which is arguably preferable. This also follows data libraries such as `pandas`.
```
Comment (rgommers) [inline apps/labs/posts/convoluted-story-behind-topk.md:141]: textual: "data libraries" -> "data science libraries" (or "dataframe libraries"?)

File: apps/labs/posts/convoluted-story-behind-topk.md [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+Despite this, the endeavour did show that there were no other compatibility concerns.
+
+### NumPy Community Discussions
+
+The discussions on the `NaN` ordering for `np.top_k` raised the issue that `np.partition` too, does not specify the `NaN` ordering. I opened a PR [gh-26716 DOC: Add clarifications np.argpartition](https://github.com/numpy/numpy/pull/26716) which clarified that `np.partition` does treat `NaN`s as the biggest element. With that merged, it is now justified to say that the current implementation of `np.top_k` follows the same sorting behaviour.
+
+However, there was still an impasse about the `NaN` ordering, so Ralf brought up the topic on my behalf at the next NumPy community meeting, which I did not join as it was very late for my time zone.
+
+Community meetings are great for these kinds of discussions since, in comparison to asynchronous conversations spread across, say, 4 threads, everything can happen in realtime.
```
Comment (rgommers) [inline apps/labs/posts/convoluted-story-behind-topk.md:177]: minor textual: "in realtime" -> "in real time"

**Review by rgommers — COMMENTED** [review]
This is a very nice blog post, thanks @JuliaPoo! The story line works quite well I think. I have only two content-related comments, and a handful of minor textual ones.

File: apps/labs/posts/convoluted-story-behind-topk.md (hunk 2 above)
Comment (JuliaPoo) [reply apps/labs/posts/convoluted-story-behind-topk.md:22]: Oh good point!

**Comment by JuliaPoo** [comment]
I've addressed the comments!

File: apps/labs/posts/convoluted-story-behind-topk.md [hunk 9]
```diff
@@ -0,0 +1,213 @@
+---
+title: 'The convoluted story Behind `np.top_k`'
```
Comment (rgommers) [inline apps/labs/posts/convoluted-story-behind-topk.md:2]:
```suggestion
title: 'The convoluted story behind `np.top_k`'
```

**Review by rgommers — APPROVED** [review]
This looks great, and is ready to go. Hitting the green button, thanks again Jules!

**Pull request merged**